Adding NOOP action

pitiwari opened the pull request:
This pr adds new action called fallthru. Below are some details and initial discussion thread along with the link for design document 

Add routes which we don't want to take any action on route match (something like fallthru), subsequently, this request will be matched with a route we want to take action on. We will add field in access log for getting metrics. There can be multiple such routes

Use cases:

This will help in getting information on these fallthru routes and query them later to see how many times we hit those particular routes. This will help in testing these routes based on route match before we enable traffic to these routes. 


below is the link for design document 

https://docs.google.com/document/d/1u_HOy96rp6WL4Y6mVqWmiO1NXySM4uvVTVXsRK9B2FY/edit

htuch commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -175,6 +175,9 @@ message Route {
     DirectResponseAction direct_response = 7;
+
+    // Return an noop action 
+    Noop  noop = 16;
```
Would `Fallthru fallthru = 16;` be clearer? It's not that route is a nop that we execute and are done, we're actually going to keep looking for the real route entry (IIRC, correct me if wrong).

htuch commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -175,6 +175,9 @@ message Route {
     // Return an arbitrary HTTP response directly, without proxying.
     DirectResponseAction direct_response = 7;
+
+    // Return an noop action 
```
Please describe what it actually does, i.e its interaction with later routing, how request processing continues, etc.

htuch commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -1038,6 +1041,13 @@ message DirectResponseAction {
+message Noop {
+      // If set to true, then when this route is hit, stream info (and hence access logs)
+      // will be populated with the name of this route. Only the NOOP routes that set this
+      // flag will be added to access logs.
```
Where in stream info is this added? Stream info is also largely an internal construct, the docs here should largely be reflecting of what's going on from an operator perspective, e.g. does this appear in access logs? How does it appear in access logs? What's the magic escape sequence to see it for text access logs, etc? I would drop reference to stream info.

pitiwari commented:
@mattklein123 @htuch  all checks are passing now. Can you start reviewing the PR ?. I will start addressing comments. Tests take 6-8 hours in completing so want do handle comments in batches

mattklein123 commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -180,6 +180,9 @@ message Route {
     // Return an arbitrary HTTP response directly, without proxying.
     DirectResponseAction direct_response = 7;
+
+    // Return an noop action
```
What does returning a noop action mean? Rephrase?

htuch replied:
This was part of my previous feedback, I would prefer this to be called `Fallthru` or something like that. @pitiwari can you address my earlier feedback and I'll take another pass, thanks.

mattklein123 commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -1062,6 +1065,13 @@ message DirectResponseAction {
 }
 
+message Noop {
+  // If set to true, then when this route is hit, stream info (and hence access logs)
```
From the user perspective, it's not clear what this actually does. Stream info as an internal concept. Is this feature generally usable without custom extensions?

mattklein123 commented on include/envoy/router/router.h:
```diff
@@ -93,6 +93,10 @@ class DirectResponseEntry : public ResponseEntry {
    */
   virtual const std::string& routeName() const PURE;
+
+  virtual bool noop() const PURE;
```
doc comments on all interface functions here and below.

mattklein123 commented on include/envoy/router/router.h:
```diff
@@ -880,7 +888,7 @@ class Config {
   virtual RouteConstSharedPtr route(const Http::HeaderMap& headers,
                                     const StreamInfo::StreamInfo& stream_info,
-                                    uint64_t random_value) const PURE;
+                                    uint64_t random_value, uint32_t&) const PURE;
```
Update doc comments, not sure what the new parameter does.

mattklein123 commented on source/common/access_log/access_log_formatter.cc:
```diff
@@ -498,6 +498,11 @@ StreamInfoFormatter::StreamInfoFormatter(const std::string& field_name) {
         return UnspecifiedValueString;
       }
     };
+  } else if (field_name == "NOOP_ROUTE_NAME") {
```
Needs documentation. Also needs implementation in gRPC access logs if we decide to go this route.

pitiwari commented:
> Thanks for working on this. I have some high level comments and questions. To be perfectly honest I'm still not sure we have the right solution here, so I would rather step back and nail the documentation. It would be very helpful for me if you could describe how you expect to use this feature from an end user perspective (including configuration) before we invest too much more time in the guts of code review. Thank you!
> 
> /wait

@mattklein123 sure i will update the design document along with more details and example confirguration. I will also take care of other comments u have

pitiwari commented:
@htuch  @mattklein123 i have addressed your comments. Updated design document as updated example as well. Please take a pass at pr

mattklein123 commented:
@pitiwari please never force push. It makes reviews much more difficult. Either merge master or add commits. Thanks!

mattklein123 commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -180,6 +180,10 @@ message Route {
     // Return an arbitrary HTTP response directly, without proxying.
     DirectResponseAction direct_response = 7;
+
+    // if action is specified as fallthru, then router will look for the next route for which action
```
nit: use proper grammar including starting sentences with capital letters, etc.

mattklein123 commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -1062,6 +1066,12 @@ message DirectResponseAction {
 }
 
+message Fallthru {
+  // If this flag is set to 1 then route name (if exists for the route) for fallthru routes will be
```
"Set to true"

Per my last comment, stream info is not a useful end user thing to describe. How is the user going to interact with this? Can you describe fully how the access log will work? This will also need to be added to the gRPC access logs in some way?

mattklein123 commented on include/envoy/router/router.h:
```diff
@@ -876,11 +896,12 @@ class Config {
    * @param headers supplies the request headers.
    * @param random_value supplies the random seed to use if a runtime choice is required. This
    *        allows stable choices between calls if desired.
+   * @param route_index specifies which route index to start looking for route on.
```
I'm not sure that the index is going to be relevant if/when the route table contains hash lookup types.

Stepping back, I'm sorry but I continue to have strong reservations that this is a general purpose feature that is useful to people. The invasiveness of this change makes me think we should look for a better way that you could handle as an extension.

What I'm thinking is:
1) What if we allows the route function to take an optional callback that asks that caller to decide if a route should be accepted?
2) You can store fall through status inside route metadata.
3) Make it so that you can write a filter that can repopulate the route using a callback function to select the final route, and do whatever logging you want along the way.

^ seems much less invasive while still allowing you to achieve your goals. @htuch WDYT?

htuch replied:
@mattklein123 are you describing something similar to the host retry picking extension approach but for routes? I kind of like where this is heading, in that we are making routing more extensible.

@alyssawilk and I have talked in the past about making routing pluggable in general; one of the challenges with a pluggable router today is that it forces you to lose the HCM/router routing logic and build it completely from scratch. I wonder if what you propose, or some variant, could be the solution here. Also CC @markdroth, we've been discussing the possibility of having a way to say "the cluster for this route comes from an extension" in an unrelated conversation.

mattklein123 replied:
> @mattklein123 are you describing something similar to the host retry picking extension approach but for routes? I kind of like where this is heading, in that we are making routing more extensible.

Yeah pretty much. I think we could do this pretty cleanly without even a new extension point (just an HTTP filter perhaps, though depending on where we want the route to be picked I suppose it might need to be a first-class extension that is runnable directly from the HCM code).

pitiwari replied:
@htuch and @mattklein123 if you both agree to the approach which @mattklein123 described above, then i can start making changes.

What if we allows the route function to take an optional callback that asks that caller to decide if a route should be accepted?
You can store fall through status inside route metadata.
Make it so that you can write a filter that can repopulate the route using a callback function to select the final route, and do whatever logging you want along the way.

mattklein123 replied:
I think this approach will work for me. @htuch ?

htuch replied:
Yeah, this sounds the right direction. @pitiwari can you put something up for review when you have the basic interfaces worked out, before going too far into writing tests etc? I'd like to see see what this looks like in code.

pitiwari replied:
@htuch sure i will do it
